recipes/self-generated-documentation.md

```diff
@@ -0,0 +1,129 @@
+- Increases adoption rate.
+- Documentation is the key to a good experience when consuming your API.
+- Reduces onboarding time.
+- Allows to test your APIs.
+- Allows to describe the structure of your APIs.
+- Allows to describe responses based on [HTTP status codes](https://httpstatuses.com/).
+- Allows to describe request body.
+
+**Swagger as example :**
+
+API documentation is a technical content deliverable.\
+This documentation contains instructions about how to effectively use and integrate with an API.\
+Demonstrating new features from Swagger allows you to:
+
+- Demonstrates the integration between API and Swagger.
+- Demonstrates new products in a client / consumer context.
+
+```shell
+  paths:
+  /products:
+    post:
+      summary: Adds a new product
+      requestBody:
+        content:
+          application/json:
+            schema:     
+              type: object
+              properties:
+                id:
+                  type: integer
+                name:
+                  type: string
+              example:   
+                id: 10
+                name: new product
+      responses:
+        '200':
+          description: OK
+```
+
+### Document effectively
+
+Most APIs are documented in OpenAPI format and are based on [Petstore sample](https://petstore.swagger.io/) provided by [swagger.io](https://swagger.io/) team. Other specifications exist: RAML and API Blueprint.
+
+- Choose your documentation standard.
+- Automate the generation of your documentation.
+
+#### How to ?
+
+We highly recommend you to choose OpenAPI standards. It's a most used standard for API documentation.\
+The OpenAPI specification was known as Swagger until version 3.0.\
+Some open source libs are available, we recommend you to choose one of them.
+We recommend to choose Hapi Swagger in case of NodeJS API. You can find the NPM package [here](https://www.npmjs.com/package/hapi-swagger).\
+Benefits:
+
+- Simple to implement and to understand for the developpers and Client.
+- Can be used with NodeJS API.
+- An active community.
+- User friendly UI.
+- Design-First : You can define an API with types and examples for each endpoint before you start implementing it.
+- You can document your APIs reponses.
+- You can document your APIs errors.
+
+```js
+import * as Hapi from '@hapi/hapi'
+import * as HapiSwagger from 'hapi-swagger'
+
+const swaggerOptions: HapiSwagger.RegisterOptions = {
+    info: {
+      title: 'Example API Documentation',
+        description: 'Some description',
+        contact: {
+          email: 'example@example.com'
+        }
+    },
+    tags: ['Example', 'API', 'Tag'],
+    basePath: `/basePath`,
+    grouping: 'tags',
+    security: [{ oidc: [] }],
+}
+
+const plugins: Array<Hapi.ServerRegisterPluginObject<any>> = [
+    {
+        plugin: Inert
+    },
+    {
+        plugin: Vision
+    },
+    {
+        plugin: HapiSwagger,
+        options: swaggerOptions
+    }
+]
+
+await server.register(plugins)
+```
+
```

Comment:
pour la partie java :

### Implementation example in Java / Kotlin / Spring

If you have a Spring application (in Java or Kotlin), you can use the [Springfox lib](https://springfox.github.io/springfox/) to:
- generate automatically an OpenAPI description of the endpoints declared in your application (json or yaml file)
- expose a Swagger UI static page on your application, displaying the generated documentation

Add Springfox dependency, for example with Gradle and SpringBoot:
```kotlin
dependencies {
    implementation "io.springfox:springfox-boot-starter:3.0.0"
}
```

Declare the API in a dedicated config file, using the `@EnableSwagger2` annotation:
```kotlin
@Configuration
@EnableSwagger2
class SwaggerConfig {
    @Bean
    fun apiDocumentation(): Docket = Docket(SWAGGER_2)
        .groupName("My API")
        .apiInfo("My API provides a great service!")
        .build()
}
```

By default, the documentation will be generated for all endpoints declared in Spring controllers.
Swagger UI is available on `<host>/swagger-ui.html`

Please refer to [Springfox documentation](http://springfox.github.io/springfox/docs/current/#quick-start-guides) or [Baeldung tutorial](https://www.baeldung.com/swagger-2-documentation-for-spring-rest-api) for further details.